proof/courier.go

```diff
@@ -1019,6 +1175,82 @@ type UniverseRpcCourier struct {
 	subscriberMtx sync.Mutex
 }
 
+// NewUniverseRpcCourier creates a new universe RPC proof courier service
+// handle.
+func NewUniverseRpcCourier(ctx context.Context, cfg *UniverseRpcCourierCfg,
+	transferLog TransferLog, localArchive Archiver, addr *url.URL,
+	lazyConnect bool) (*UniverseRpcCourier, error) {
+
+	courier := UniverseRpcCourier{
+		cfg:           cfg,
+		addr:          addr,
+		localArchive:  localArchive,
+		backoffHandle: NewBackoffHandler(cfg.BackoffCfg, transferLog),
+		subscribers:   make(map[uint64]*fn.EventReceiver[fn.Event]),
+	}
+
+	// If we're not lazy connecting, then we'll attempt to connect to the
+	// courier service immediately.
+	if !lazyConnect {
+		err := courier.ensureConnect(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("unable to connect to courier "+
+				"service during courier handle "+
+				"instantiation: %w", err)
+		}
+	}
+
+	return &courier, nil
+}
+
+// ensureConnect ensures that the courier handle is connected to the remote
+// courier service.
+//
+// This method does nothing if a service connection is already established.
+func (c *UniverseRpcCourier) ensureConnect(ctx context.Context) error {
+	// If we're already connected, we'll return early.
+	if c.rawConn != nil && c.client != nil {
+		// Check the gRPC connection state to determine if the
+		// connection is ready.
+		gRpcConnState := c.rawConn.GetState()
+
+		connStatus, err := NewCourierConnStatusFromRpcStatus(
+			gRpcConnState,
+		)
+		if err != nil {
+			return fmt.Errorf("universe RPC courier unable to "+
+				"determine connection status: %w", err)
+		}
+
+		if connStatus.IsPending() {
+			return nil
+		}
+	}
+
+	// At this point, we know that the connection is not ready. We'll now
+	// attempt to establish a new connection to the courier service.
+	dialOpts, err := serverDialOpts()
```

Comment:
You mentioned offline that you wanted the option for the connection setup to be blocking; I think adding the `WithBlock()` `grpc.DialOption` would work for that.

https://pkg.go.dev/google.golang.org/grpc@v1.59.0#WithBlock

Reply:
That's what I've been playing around with in conjunction with `DialContext`